pkg/controller/egress/controller.go

```diff
@@ -92,17 +157,109 @@ func (c *EgressController) Run(stopCh <-chan struct{}) {
 	klog.Infof("Starting %s", controllerName)
 	defer klog.Infof("Shutting down %s", controllerName)
 
-	cacheSyncs := []cache.InformerSynced{c.egressListerSynced, c.groupingInterfaceSynced}
+	cacheSyncs := []cache.InformerSynced{c.egressListerSynced, c.externalIPPoolListerSynced, c.groupingInterfaceSynced}
 	if !cache.WaitForNamedCacheSync(controllerName, stopCh, cacheSyncs...) {
 		return
 	}
 
+	// Initialize the ipAllocatorMap and ipAllocationMap with the existing ExternalIPPools and Egresses.
+	ipPools, _ := c.externalIPPoolLister.List(labels.Everything())
+	for _, ipPool := range ipPools {
+		c.createOrUpdateIPAllocator(ipPool)
+	}
+	egresses, _ := c.egressLister.List(labels.Everything())
+	for _, egress := range egresses {
+		c.updateIPAllocation(egress)
+	}
+
 	for i := 0; i < defaultWorkers; i++ {
 		go wait.Until(c.egressGroupWorker, time.Second, stopCh)
 	}
 	<-stopCh
 }
 
+// updateIPAllocation sets the EgressIP of an Egress as allocated in the specified ExternalIPPool and records the
+// allocation in ipAllocationMap.
+func (c *EgressController) updateIPAllocation(egress *egressv1alpha2.Egress) {
+	// Ignore Egress that is not associated to ExternalIPPool or doesn't have EgressIP assigned.
+	if egress.Spec.ExternalIPPool == "" || egress.Spec.EgressIP == "" {
+		return
+	}
+	ipAllocator, exists := c.getIPAllocator(egress.Spec.ExternalIPPool)
+	if !exists {
+		klog.ErrorS(externalIPPoolNotFound, "Failed to allocate EgressIP", "egress", egress.Name, "ip", egress.Spec.EgressIP, "pool", egress.Spec.ExternalIPPool)
+		return
+	}
+	ip := net.ParseIP(egress.Spec.EgressIP)
+	err := ipAllocator.AllocateIP(ip)
+	if err != nil {
+		klog.ErrorS(err, "Failed to allocate EgressIP", "egress", egress.Name, "ip", egress.Spec.EgressIP, "pool", egress.Spec.ExternalIPPool)
+		return
+	}
+	// Record the valid IP allocation.
+	c.setIPAllocation(egress.Name, ip, egress.Spec.ExternalIPPool)
+	klog.InfoS("Allocated EgressIP", "egress", egress.Name, "ip", egress.Spec.EgressIP, "pool", egress.Spec.ExternalIPPool)
+}
+
+// createOrUpdateIPAllocator creates or updates the IP allocator based on the provided ExternalIPPool.
+// Currently it's assumed that only new ranges will be added and existing ranges should not be deleted.
+// TODO: Use validation webhook to ensure it.
+func (c *EgressController) createOrUpdateIPAllocator(ipPool *egressv1alpha2.ExternalIPPool) bool {
+	changed := false
+	c.ipAllocationMutex.Lock()
+	defer c.ipAllocationMutex.Unlock()
+
+	existingIPRanges := sets.NewString()
+	multiIPAllocator, exists := c.ipAllocatorMap[ipPool.Name]
```

Comment:
I noticed you used the `ipAllocationMutex` mutex here and below, when you are in fact accessing `ipAllocatorMap`. Is that a typo?

Reply:
Definitely a typo, relied on auto-completion too much. Thanks for catching it!